compiler/passes/src/static_single_assignment/rename_expression.rs

```diff
@@ -169,13 +171,46 @@ impl ExpressionConsumer for StaticSingleAssigner {
             })
             .collect();
 
+        // Reorder the members to match that of the struct definition.
+
+        // Lookup the struct definition.
+        // Note that type checking guarantees that the correct struct definition exists.
+        let struct_definition: &Struct = self.symbol_table.borrow().structs.get(&input.name.name).unwrap();
+
+        // Initialize the list of reordered members.
+        let mut reordered_members = Vec::with_capacity(members.len());
+
+        // Collect the members of the init expression into a map.
+        let mut member_map: IndexMap<Symbol, StructVariableInitializer> = members
+            .into_iter()
+            .map(|member| (member.identifier.name, member))
+            .collect();
+
+        // If we are initializing a record, add the `owner` and `gates` fields, first and second respectively.
+        // Note that type checking guarantees that the above fields exist.
+        if struct_definition.is_record {
+            // Add the `owner` field.
+            reordered_members.push(member_map.remove(&sym::owner).unwrap());
+            // Add the `gates` field.
+            reordered_members.push(member_map.remove(&sym::gates).unwrap());
+        }
+
+        // For each member of the struct definition, push the corresponding member of the init expression.
+        for member in &struct_definition.members {
+            // If the member is part of a record and it is `owner` or `gates`, then we have already added it.
+            if !(struct_definition.is_record && matches!(member.identifier.name, sym::owner | sym::gates)) {
+                // Lookup and push the member of the init expression.
+                reordered_members.push(member_map.remove(&member.identifier.name).unwrap());
```

Comment:
At this point type checking has guaranteed that the member will be found, i.e. that `unwrap` won't fail, right? (Just double-checking.)

Reply:
Yes, that's right.
I'll make a note of the `unwrap`.